### PR TITLE
test backward() calls don't hit the CPU fallback

### PR DESCRIPTION
**Pull request by bdhirsh**
This should help CI catch errors where a derivative formula is changed in core to use an op that pytorch/xla doesn't have a lowering for, which silently causes usages of that op during autograd to hurt performance. (There was a specific example recently with `abs`'s derivative being updated to use `sgn` instead of `sin`).

I added an example of what we could add to each op's test, for `TestAbs` and for `TestSign`. Both tests now also run `.backward()`, assert that the accumulated gradients match, and assert that no operators hit the CPU fallback.

The test correctly fails for `TestAbs`, because `sgn` doesn't have a lowering yet! So we can wait to land that once https://github.com/pytorch/xla/pull/3045 lands.

### CODE REVIEW

File: test/cpp/test_aten_xla_tensor.cpp [hunk 1]
```diff
@@ -3014,13 +3014,15 @@ TEST_F(AtenXlaTensorTest, TestBitwiseNotInPlace) {
 
 TEST_F(AtenXlaTensorTest, TestSign) {
   torch::Tensor a =
-      torch::randn({2, 2}, torch::TensorOptions(torch::kFloat)) * 100.0;
+      torch::randn({200, 200}, torch::TensorOptions(torch::kFloat).requires_grad(true));
```
Comment (bdhirsh) [inline test/cpp/test_aten_xla_tensor.cpp:3017]: one annoying thing - the test breaks with an autograd error if we do `a = torch.tensor(..., requires_grad=true) * 100`. I'm not totally sure what the autograd interaction is, but it looks like the reference cpu tensor needs to be created in one shot. Maybe we could avoid that by doing `.detach().set_requires_grad(true)`, but that felt clunkier.

File: test/cpp/cpp_test_util.cpp [hunk 2]
```diff
@@ -78,8 +78,15 @@ at::Tensor ToCpuTensor(const at::Tensor& tensor) {
...
 
 torch::Tensor CopyToDevice(const torch::Tensor& tensor,
-                           const torch::Device& device) {
-  return tensor.to(device, /*non_blocking=*/false, /*copy=*/true);
+                           const torch::Device& device,
+                           bool requires_grad) {
+  auto output = tensor.to(device, /*non_blocking=*/false, /*copy=*/true);
+  if (requires_grad) {
+    // The .detach().requires_grad_(true) is to make sure that the new tensor is a leaf in the autograd graph.
+    return output.detach().set_requires_grad(true);
```
Comment (bdhirsh) [inline test/cpp/cpp_test_util.cpp:86]: the detach().set_requires_grad(true) is also to appease autograd. We need the xla tensor "input" for each test to be a "leaf node", from autograd's perspective.

File: test/cpp/test_aten_xla_tensor.cpp (hunk 1 above)
Comment (JackCaoG) [reply test/cpp/test_aten_xla_tensor.cpp]: I think the purpose of *100 is to get the random number within (-100, 100) instead of (-1, 1). Would do a inplace mul fix the problem here?

Comment (bdhirsh) [reply test/cpp/test_aten_xla_tensor.cpp]: I just read the test way too quickly 😛 yeah, that should probably work. I'll fix it

Comment (bdhirsh) [reply test/cpp/test_aten_xla_tensor.cpp]: fixed, thanks. I had to set `requires_grad=true` after the inplace update, since autograd doesn't handle leaf mutations.

File: test/cpp/cpp_test_util.cpp (hunk 2 above)
Comment (JackCaoG) [reply test/cpp/cpp_test_util.cpp:86]: so... when copying tensor using `tensor.to`, will `require_grad` field also being copied over? I am guessing you need to reset `set_requires_grad` because you called `deteach` but want to double check. 😄

Comment (bdhirsh) [reply test/cpp/cpp_test_util.cpp:86]: didn't know until I made this PR, but yep 😛 

```
device = xm.xla_device()
a = torch.ones(2, 2, device='cpu', requires_grad=True)
b = a.to(device)
print(b.requires_grad) # printed "True"
```

File: test/cpp/cpp_test_util.cpp [hunk 3]
```diff
@@ -278,6 +285,15 @@ std::vector<at::Tensor> ExecuteAndFetch(absl::Span<const ir::Value> roots,
   return Fetch(results);
 }
 
+void AssertBackward(const torch::Tensor& xla_output,
+                    const torch::Tensor& xla_input,
```
Comment (JackCaoG) [inline test/cpp/cpp_test_util.cpp:289]: Can we generalize this to take `vector<torch::Tensor>`? For some op we take more than one input.

**Review by JackCaoG — CHANGES_REQUESTED** [review]

File: test/cpp/cpp_test_util.cpp (hunk 3 above)
Comment (bdhirsh) [reply test/cpp/cpp_test_util.cpp]: done!

**Review by JackCaoG — APPROVED** [review]
I think you need to rebase to pass the CPU python CI but otherwise LGTM
